lib/go/test/flow_epoch_test.go

```diff
@@ -1516,3 +1517,203 @@ func TestEpochReset(t *testing.T) {
 		assertEqual(t, CadenceUFix64("249999.99300000"), result)
 	})
 }
+
+func TestEpochRecover(t *testing.T) {
```

Comment:
Could you add documentation for the test here?

Currently, I think we are testing:
- execute an epoch recover where `recoveryEpochCounter == currentEpochCounter+1`
- execute an epoch recover where `recoveryEpochCounter == currentEpochCounter`

After we explicitly include the `recoveryEpochCounter`, I would also suggest testing the case where:
- `recoveryEpochCounter != currentEpochCounter+1` and the `unsafeAllowOverwrite` flag is not set

Reply:
https://github.com/onflow/flow-core-contracts/pull/420/commits/1e41ecf0ca8519eec08751639a3ad3928072c8a0

Reply:
Not sure if I follow the last point, if we have 2 separate funcs one to recover new epoch and one to recover current epoch we dont need to unsafeAllowOverwrite we dont pass unsafeAllowOverwrite to any of the FlowEpoch funcs we only use it in the transaction to decide which func `recoverNewEpoch` or `recoverCurrentEpoch`   to call.

Reply:
Once we are passing in the `recoveryEpochCounter`, we must use that value as the epoch counter for the recovery epoch 100% of the time, otherwise we will produce an incompatible recovery epoch state. `recoveryEpochCounter` is the source of truth.

In the commit which was merged to the feature branch, the logic is:
1. if `unsafeAllowOverwrite` is set, then use `recoveryEpochCounter` as the counter for the recovery epoch
2. otherwise, ignore the `recoveryEpochCounter` input and assume that the the smart contract and protocol state epoch counters are synchronized

I think conditional branch 2 is unsafe and we should respect the `recoveryEpochCounter` codepath in all cases. Sorry if I didn't explain this properly in the previous review. The purpose of including `recoveryEpochCounter` at all is to make sure it is used as the epoch counter for the recovery epoch being constructed in the transaction, in all circumstances.